tools/packages/wesl/src/LinkedWesl.ts

```diff
@@ -0,0 +1,171 @@
+import { SrcMap } from "mini-parse";
+import type { WeslDevice } from "./WeslDevice";
+import { offsetToLineNumber } from "./Util";
+import { assertThat } from "../../mini-parse/src/Assertions";
+
+/** Results of shader compilation. Has {@link WeslGPUCompilationMessage} which are aware of the WESL module that an error was thrown from. */
+export interface WeslGPUCompilationInfo extends GPUCompilationInfo {
+  messages: WeslGPUCompilationMessage[];
+}
+
+export interface WeslGPUCompilationMessage extends GPUCompilationMessage {
+  module: {
+    url: string; // LATER this should be a qualified module path. And something else should map it to a URL that is relative to the correct place.
+    text?: string; // LATER: I don't think that the text should be a part of the compilation message. Instead the module url should be usable as a key.
+  };
+}
+
+/** A {@link GPUValidationError} with an inner error (for a stack trace). Can also point at a WESL source file. */
+export class ExtendedGPUValidationError extends GPUValidationError {
+  public cause?: Error;
+  public compilationInfo?: WeslGPUCompilationInfo;
+  constructor(message: string, options?: { cause?: Error }) {
+    super(message);
+    this.cause = options?.cause;
+  }
+}
+
+/**
+ * Multiple WESL files that have been linked together to produce WGSL code.
+ *
+ * Call {@link LinkedWesl.createShaderModule} on a {@link WeslDevice} to make the error reporting aware of the WESL code.
+ */
+export class LinkedWesl {
+  constructor(public sourceMap: SrcMap) {}
+
+  /**
+   * Creates a {@link GPUShaderModule}.
+   *
+   * The compilation info {@link GPUShaderModule.getCompilationInfo} can be remapped with {@link mapGPUCompilationInfo}
+   * @param device GPUDevice. Preferably a {@link WeslDevice} for better error reporting.
+   * @param descriptor - Description of the {@link GPUShaderModule} to create.
+   */
+  createShaderModule(
+    device: GPUDevice | WeslDevice,
+    descriptor: Omit<GPUShaderModuleDescriptor, "code">,
+  ): GPUShaderModule {
```

Comment:
Our custom `createShaderModule`, which does the whole "catch the underlying error, remap it to WESL, and rethrow it"